### ISSUE TITLE
Training stuck with custom loss

### ISSUE THREAD

**Issue by pjsjongsung** [tensorflow/tensorflow#50083]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): pip
- TensorFlow version (use command below): 2.5.0
- Python version: 3.8.0
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory: RTX 2070 Ti

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
The training stops running after printing Epoch 1/20 when I use model.fit to run my model. It was fine until I added this loss function:

```
class snc_loss(tf.keras.losses.Loss):
    def call(self, y_true, y_pred):
        l_sigma_I = tf.constant(100.0, dtype=tf.float32)
        l_sigma_X = tf.constant(16.0, dtype=tf.float32)
        l_R = tf.constant(5.0, dtype=tf.float32)
        shape = tf.shape(y_pred)
        loss = tf.ones((shape[0],)) * tf.cast(shape[3], dtype=tf.float32)
        for channel in tf.range(shape[-1]):
            sub = tf.zeros((shape[0],))
            denom = tf.ones((shape[0],)) * 1e-6
            for i in tf.range(shape[1] * shape[2]):
                y_i = i%shape[2]
                x_i = i//shape[2]
                for j in tf.range(i+1, shape[1]*shape[2]):
                    y_j = j%shape[2]
                    x_j = j//shape[2]
                    eu_dis = tf.cast(tf.math.square(x_i - x_j)+tf.math.square(y_i - y_j), dtype=tf.float32)
                    pixel_dis = tf.cast(tf.math.square(y_true[:, x_i, y_i] - y_true[:, x_j, y_j]), dtype=tf.float32)
                    if tf.math.less(tf.math.sqrt(eu_dis), l_R):
                        weight = tf.math.exp(-1 * pixel_dis / l_sigma_I - eu_dis / l_sigma_X)
                        sub += weight * y_pred[:, x_j, y_j, channel] * y_pred[:, x_i, y_i, channel]
                        denom += weight * y_pred[:, x_i, y_i, channel]

            loss -= sub/denom

        return loss
```

**Describe the expected behavior**
It should be training

**[Contributing](https://www.tensorflow.org/community/contribute)** - Do you
want to contribute a PR? (yes/no): - Briefly describe your candidate solution
(if contributing): no

**Standalone code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

https://colab.research.google.com/drive/1CqZTxlyUNlUDkNipoVRPhA3HHzjlvuUj?usp=sharing

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

The error was still there when I was using Tensorflow 2.4.1, but it used to tell me SubProcess ended with return code : 0
Also, when I run a similar code as a job on a server, it throws a bus error, so might be related to memory.

**Comment by tilakrayal** [comment]
I was able to reproduce the issue in tf v2.4,v2.5 and nightly.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/2bf52ce28c35b618568956ac8f384921/50083.ipynb).

**Comment by Saduf2019** [comment]
@pjsjongsung 
You should define your constants in init, You are creating constants at every step is , there is a lot of complexity in your function.
please try to simplify your function with tf ops and broadcasting concepts instead of the python loops,The training does proceed but the problem is with the loss which is taking too long.

Kindly open a stackoverflow/tf discussion forum issue as this is not a bug or feat request and move this to closed status.

**Comment by pjsjongsung** [comment]
I am using loops because of a memory issue and cannot initialize weights because then the matrix is too big. But I get the constant issue. Closing this. Thank you!
